### ISSUE TITLE
Enable support for Multi-Shard queries with Entity Framework

### ISSUE THREAD

**Issue by SilviaDoomra** [Azure/elastic-db-tools#61]

**Comment by abatishchev** [comment]
Any news on this, please? Currently this seems to be the only roadblockerer left for us on the way to leverage Ealasric Sharding in full.

**Comment by jaredmoo** [comment]
Can you explain what the scenario is that you need? This will allow us to design appropriately. Thanks.

**Comment by abatishchev** [comment]
For now we're having a single sql db with Entity Framework 6 DbContext, and OData exposed over Web API on top of it.
And we're going to shard the db by tenant id. So would be nice to be able to run multi-shard queries transparently.
Otherwise Web API OData endpoint (e.g, `/data/earnings?$filter=Id eq 1`) should require mandatory parameter tenantId (e.g. `/data/earnings?tenantId=5&$filter=Id eq 1`) and API consumers to be aware of sharding and tenant ids.

**Comment by jaredmoo** [comment]
Ok, I see. Do you have a lot of APIs that you want to be cross-tenant? In most elastic scale applications, the API exposed by the application only allows the caller to access their own individual tenant id.

In theory I think it should be possible to write an adapter for EF to use `MultiShardConnection` and so on. I'd be happy to discuss a design with you and review pull requests, but I won't be available in the foreseeable future to do dev/test work on this myself.

Have you considered using [SQL elastic query?](https://azure.microsoft.com/en-us/documentation/articles/sql-database-elastic-query-overview/) This is a transparent wrapper around MultiShardQuery built into Azure SQL DB, perhaps it could help in your scenario.

**Comment by abatishchev** [comment]
Our application consists from Web API, heavy compute using WebJobs and SQL Azure db. Web API calls under few categories: data ingest, infrastructure and reporting.
For the former two it's not a problem to make them tenant/sharding-aware, but for the later one almost the whole purpose of it is to be tenant/sharding-unaware.
Of course it should be possible to workaround this around (in a way such as I've mentioned earlier) but I'd like to try fiest to implement it in a transparent manner.

Contributing a pull request sounds perfect to me! The willingness to review one is a half-way to successes :) Aso embrace One Microsoft, you know.
Since `MultiShardConnection.CreateCommand()` returns `DbConnection` it should be possible to pass it to a ctor of `DbContext`. Then `DbContext.DbSet<T>` returns `IQuerable<T>` yo consume by OData, etc.
Are we on the same page here, or you have a different idea from the top of your head?

P.S. I don't think Elastic Queries would work for us since the DAL is completely abstracted in our case, hidden by EF/OData and we never write/see/accesses the resulting SQL. Will definitely look into this option if others won't work.

**Comment by jaredmoo** [comment]
Yes I was thinking of something like this, the first problem that I can think of is that `MultiShardConnection` actually does not implement `DbConnection` - this is intentionally because `MultiShardConnection` doesn't implement `Open`/`Close` since it internally opens and closes connections at the time of command execution.

**Comment by jaredmoo** [comment]
This can be worked around, e.g. you could create a wrapper around `MultiSharedConnection` that implements `DbConnection` with no-op implementations of `Open`, `Close`, etc.

**Comment by abatishchev** [comment]
I forked the repo and added a draft for [MultiShardConnectionAdapter](https://github.com/abatishchev/elastic-db-tools/blob/mshconn-adapter-1/Src/ElasticScale.Client/Query/MultiShardConnectionAdapter.cs).

**Comment by abatishchev** [comment]
hey @jaredmoo, please take a look and give some pointers where to go next. thanks!

**Comment by jaredmoo** [comment]
Does it work when you pass it to a DbContext? I suggest writing unit tests and see what happens :)

**Comment by abatishchev** [comment]
Oh, sorry for the late response. You're right, it doesn't work as it currently is. But it doesn't work either with all fixes I tried to apply and didn't push yet. It works as a stand-alone connection but integration with EF6 requires a ton of extra changes, such as `DbProviderFactory` and `DbProviderServices`.

**Comment by stevozilik** [comment]
Any progress on this? It's a blocker for us as well
